### PR TITLE
feat: parse system properties as string, closes MISC-388

### PR DESCRIPTION
**Pull request by ccedric**
feat: Parse sytem properties as string, closes MISC-388

Motivation:
- multiple build plugins have trouble parsing system properties to a Map

Modifications:
- added a PropertiesParserUtil which parse a string of properties to a map
- following format is used: key1=value1,key2=value2

--------------

feat: Handle environment variables, closes MISC-368

Motivation:
- the public API allows passing extra env variables

Modifications:
- accepts env variables as a Map

### CODE REVIEW

File: src/main/java/io/gatling/plugin/util/PropertiesParserUtil.java [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+        .collect(Collectors.toMap(property -> property[0].trim(), property -> property[1].trim()));
+  }
+
+  /**
+   * Returns the correct properties computed with the map and the encoded string
+   * @param propertiesMap Properties map
+   * @param propertiesString Properties encoded into a String
+   * @return Returns the content of the properties string if not empty, returns the properties map otherwise
+   */
+  public static Map<String, String> getProperties(
```
Comment (ccedric) [inline src/main/java/io/gatling/plugin/util/PropertiesParserUtil.java:48]: I'm not sure if the string should overrides the map, or just merge with it.
The string will mainly be used in command line (for a specific run), whereas the map will be used in the config line (for the whole project).

**Review by tpetillot — CHANGES_REQUESTED** [review]

File: src/main/java/io/gatling/plugin/BatchEnterprisePluginClient.java [hunk 2]
```diff
@@ -58,10 +59,15 @@ public long uploadPackageWithSimulationId(UUID simulationId, File file)
 
   @Override
   public SimulationStartResult uploadPackageAndStartSimulation(
-      UUID simulationId, Map<String, String> systemProperties, String simulationClass, File file)
+      UUID simulationId,
+      Map<String, String> systemProperties,
```
[the remark this answers is not in the document]
Comment (ccedric) [reply src/main/java/io/gatling/plugin/BatchEnterprisePluginClient.java:63]: I think we should keep as much logic as possible in the commons.
It includes the choice of overriding the map by the formatted string for example.
It does add more params to the public functions, but I think we don't really care as it is a private project.

File: src/main/java/io/gatling/plugin/util/PropertiesParserUtil.java [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+   * @return Returns the content of the properties string if not empty, returns the properties map *
+   *     otherwise
+   */
+  public static Map<String, String> getProperties(Properties properties) {
+    Map<String, String> parsedProperties = parseProperties(properties.propertiesString);
+    if (!parsedProperties.isEmpty()) {
+      return parsedProperties;
+    }
+    return properties.propertiesMap;
+  }
```
Comment (tpetillot) [inline src/main/java/io/gatling/plugin/util/PropertiesParserUtil.java:57]: Shouldn't that be defined on `Properties`? More over, what's the point to keep both in memory if only one will be consumed.

IMO, this utility class should be able to define properties from string, maybe if really needed (but it should be the responsability of build tools IMO) the override mechanism.

Properties should still be only a map.

File: src/main/java/io/gatling/plugin/util/PropertiesParserUtil.java [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+   * @return The map containing the properties, minus malformed properties
+   */
+  protected static Map<String, String> parseProperties(String properties) {
+    if (properties == null) {
+      return Collections.emptyMap();
+    }
+    return Arrays.stream(properties.split(","))
+        .map(property -> property.split("="))
+        .filter(property -> property.length == 2)
+        .collect(Collectors.toMap(property -> property[0].trim(), property -> property[1].trim()));
```
Comment (Isammoc) [inline src/main/java/io/gatling/plugin/util/PropertiesParserUtil.java:40]: **blocking:** Potential issues:

```java

  @Test
  void parseProperties_WhitespacePropertyName() {
    String properties = "key1 name=value1,key2=value2";
    final Map<String, String> propertiesMap = new HashMap<>();
    propertiesMap.put("key2", "value2");
    assertEquals(propertiesMap, PropertiesParserUtil.parseProperties(properties));
  }

  @Test
  void parseProperties_escapedComma() {
    String properties = "key1=value1.1\\,value1.2,key2=value2";
    final Map<String, String> propertiesMap = new HashMap<>();
    propertiesMap.put("key1", "value1.1,value1.2");
    propertiesMap.put("key2", "value2");
    assertEquals(propertiesMap, PropertiesParserUtil.parseProperties(properties));
  }

  @Test
  void parseProperties_equalSignInContent() {
    String properties = "key1=v1=value1,key2=value2";
    final Map<String, String> propertiesMap = new HashMap<>();
    propertiesMap.put("key1", "v1=value1");
    propertiesMap.put("key2", "value2");
    assertEquals(propertiesMap, PropertiesParserUtil.parseProperties(properties));
  }
```

The last one can easily be done:

```suggestion
    return Arrays.stream(properties.split(","))
        .map(property -> property.split("=", 2))
        .filter(property -> property.length == 2)
        .collect(Collectors.toMap(property -> property[0].trim(), property -> property[1].trim()));
```

Comment (slandelle) [reply src/main/java/io/gatling/plugin/util/PropertiesParserUtil.java:40]: No need to try to be perfect on this. The usage is System props and env vars. We don't have to support any weird data psychos might come up with. If weird data doesn't work, so be it.

**Review by slandelle — APPROVED** [review]

**Review by guilgaly — APPROVED** [review]

File: src/main/java/io/gatling/plugin/util/PropertiesParserUtil.java (hunk 4 above)
Comment (Isammoc) [reply src/main/java/io/gatling/plugin/util/PropertiesParserUtil.java:40]: 1. It's not weird to have a `=` or `,` in a wanted value for the simulation
2. whitespace in property name or in environment variable name can be a source of security issue.
(name="echo 'PWN'; REAL_ENV")

We might execute any arbitrary command before launching the run.

Comment (slandelle) [reply src/main/java/io/gatling/plugin/util/PropertiesParserUtil.java:40]: I'd say to throw an exception if there's a whitespace in the key.

Comment (Isammoc) [reply src/main/java/io/gatling/plugin/util/PropertiesParserUtil.java:40]: So, quick wins:
* Add the `, 2` for split limit (to allow `=` in value)
* Throw an exception if there's whitespace

The escape of the comma `,` can be cumbersome to overcome

:ok_hand:

Comment (slandelle) [reply src/main/java/io/gatling/plugin/util/PropertiesParserUtil.java:40]: Fine with me. @ccedric could you please change accordingly?

**Comment by notdryft** [comment]
There a some typos to fix in the commit message

**Comment by notdryft** [comment]
> feat: Parse sytem
> feat: Parse sy**s**tem

> added a PropertiesParserUtil which parse
> added a PropertiesParserUtil which parse**s**

File: src/main/java/io/gatling/plugin/util/PropertiesParserUtil.java [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+      return Collections.emptyMap();
+    }
+    return Arrays.stream(properties.split(","))
+        .map(property -> property.split("=", 2))
+        .filter(property -> property.length == 2)
+        .collect(
+            Collectors.toMap(
+                property -> {
+                  String key = property[0].trim();
+                  if (key.contains(" ")) {
```
Comment (Isammoc) [inline src/main/java/io/gatling/plugin/util/PropertiesParserUtil.java:44]: Space is not the only problematic character

```suggestion
                  if (key.matches(".*\\s.*")) {
```

Note: `\\s` will match ANY whitespace character from UTF-8

**Review by bastien-gatling — APPROVED** [review]

**Review by Isammoc — APPROVED** [review]

**Pull request merged**